### PR TITLE
size_unit and correlation matrix

### PR DESCRIPTION
**Pull request by IKrukov-HORIS**

### CODE REVIEW

File: python-package/lets_plot/bistro/corr.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        self.data = data
+        self.show_legend = show_legend
+        self.format = format if format else '.2f'
+        self.reverse_y = flip if flip else False
+        self.text_color = None
+        self.tiles_layer = None
+        self.points_layer = None
+        self.labels_layer = None
+
+    def get_format(self, format):
```
Comment (alshan) [inline python-package/lets_plot/bistro/corr.py:54]: Use one leading underscore for non-public methods and instance variables.

File: python-package/lets_plot/bistro/corr.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+def reverse_type(type):
+    if type == 'upper':
+        return 'lower'
+    elif type == 'lower':
+        return 'upper'
+
+    return type
+
+
+class corr_plot_builder:
```
Comment (alshan) [inline python-package/lets_plot/bistro/corr.py:42]: Add doc-string  to class and to public members.

File: python-package/lets_plot/bistro/corr.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+from lets_plot.plot.scale import scale_y_discrete_reversed, scale_color_gradient2
+from lets_plot.plot.scale_identity import scale_size_identity
+from lets_plot.plot.coord import coord_fixed
+from lets_plot.plot.theme_ import theme, element_blank
+from lets_plot.plot.tooltip import layer_tooltips
+
+__all__ = ['corr_plot_builder', 'corr_plot']
+
+
+def add_common_params(plot, reverse_y):
```
Comment (alshan) [inline python-package/lets_plot/bistro/corr.py:18]: Let's use leading underscore in name to indicate these globals are "module non-public".

File: plot-base-portable/src/commonMain/kotlin/jetbrains/datalore/plot/base/geom/PointGeom.kt [hunk 4]
```diff
@@ -47,18 +49,40 @@ open class PointGeom : GeomBase() {
 
             if (SeriesUtil.allFinite(x, y)) {
                 val location = helper.toClient(DoubleVector(x!!, y!!), p)
-
                 val shape = p.shape()!!
-                targetCollector.addPoint(i, location, shape.size(p) / 2,
+                val scale = getScale(ctx, p)
```
Comment (alshan) [inline plot-base-portable/src/commonMain/kotlin/jetbrains/datalore/plot/base/geom/PointGeom.kt:53]: The name 'scale' is overused. Let's use something more concrete. What is this 'scale' anyways? Why it's depending on 'p'?

File: plot-base-portable/src/commonMain/kotlin/jetbrains/datalore/plot/base/geom/TextGeom.kt [hunk 5]
```diff
@@ -20,19 +18,28 @@ import jetbrains.datalore.plot.common.data.SeriesUtil
 class TextGeom : GeomBase() {
     var formatter: NumberFormat? = null
     var naValue = DEF_NA_VALUE
+    var sizeUnit: String? = null
```
Comment (alshan) [inline plot-base-portable/src/commonMain/kotlin/jetbrains/datalore/plot/base/geom/TextGeom.kt:21]: IMO all 'var'-s must be private.

File: plot-base-portable/src/commonMain/kotlin/jetbrains/datalore/plot/base/geom/util/GeomHelper.kt [hunk 6]
```diff
@@ -5,15 +5,14 @@
...
 import jetbrains.datalore.base.geometry.DoubleRectangle
 import jetbrains.datalore.base.geometry.DoubleVector
-import jetbrains.datalore.plot.base.CoordinateSystem
-import jetbrains.datalore.plot.base.DataPointAesthetics
-import jetbrains.datalore.plot.base.GeomContext
-import jetbrains.datalore.plot.base.PositionAdjustment
+import jetbrains.datalore.plot.base.*
 import jetbrains.datalore.plot.base.aes.AesScaling
 import jetbrains.datalore.plot.base.aes.AestheticsUtil
+import jetbrains.datalore.plot.base.geom.PointGeom
```
Comment (alshan) [inline plot-base-portable/src/commonMain/kotlin/jetbrains/datalore/plot/base/geom/util/GeomHelper.kt:15]: Is this dependency necessary?

File: plot-base-portable/src/commonMain/kotlin/jetbrains/datalore/plot/base/geom/util/GeomHelper.kt [hunk 7]
```diff
@@ -255,5 +254,23 @@ open class GeomHelper(private val myPos: PositionAdjustment, coord: CoordinateSy
...
+        const val X = "x"
+        const val Y = "y"
+
+        fun getSizeUnitAes( sizeUnitName: String ) : Aes<Double> {
+            Preconditions.checkArgument(
+                sizeUnitName == X || sizeUnitName == Y,
+                "size_unit value must be either '${X}' or '${Y}'"
+            )
+
+            if ( sizeUnitName == X) {
```
Comment (alshan) [inline plot-base-portable/src/commonMain/kotlin/jetbrains/datalore/plot/base/geom/util/GeomHelper.kt:267]: Re-write like:
```
when (sizeUnitName.toLowerCase()) {
  "x" -> 
  "y" ->
  else -> error
```

File: plot-base-portable/src/commonMain/kotlin/jetbrains/datalore/plot/base/geom/util/GeomHelper.kt [hunk 8]
```diff
@@ -255,5 +254,23 @@ open class GeomHelper(private val myPos: PositionAdjustment, coord: CoordinateSy
             shape.setStroke(stroke, strokeAlpha)
             shape.setStrokeWidth(AesScaling.strokeWidth(p))
         }
+
+        const val X = "x"
```
Comment (alshan) [inline plot-base-portable/src/commonMain/kotlin/jetbrains/datalore/plot/base/geom/util/GeomHelper.kt:258]: remove

File: plot-base-portable/src/commonMain/kotlin/jetbrains/datalore/plot/base/geom/util/GeomHelper.kt [hunk 9]
```diff
@@ -255,5 +254,23 @@ open class GeomHelper(private val myPos: PositionAdjustment, coord: CoordinateSy
...
+        const val Y = "y"
+
+        fun getSizeUnitAes( sizeUnitName: String ) : Aes<Double> {
+            Preconditions.checkArgument(
```
Comment (alshan) [inline plot-base-portable/src/commonMain/kotlin/jetbrains/datalore/plot/base/geom/util/GeomHelper.kt:262]: remove

File: plot-base-portable/src/commonMain/kotlin/jetbrains/datalore/plot/base/geom/util/GeomHelper.kt [hunk 10]
```diff
@@ -255,5 +254,23 @@ open class GeomHelper(private val myPos: PositionAdjustment, coord: CoordinateSy
...
+                "size_unit value must be either '${X}' or '${Y}'"
+            )
+
+            if ( sizeUnitName == X) {
+                return Aes.X
+            } else if ( sizeUnitName == Y) {
+                return Aes.Y
+            }
+
+            error("Wrong size_unit value")
```
Comment (alshan) [inline plot-base-portable/src/commonMain/kotlin/jetbrains/datalore/plot/base/geom/util/GeomHelper.kt:273]: more descriptive message.

File: plot-config-portable/src/commonMain/kotlin/jetbrains/datalore/plot/config/GeomProtoClientSide.kt [hunk 11]
```diff
@@ -118,6 +118,19 @@ class GeomProtoClientSide(geomKind: GeomKind) : GeomProto(geomKind) {
                 if (opts.has(Point.ANIMATION)) {
                     geom.animation = opts[Point.ANIMATION]
                 }
+
+                if (opts.has(Point.SIZE_UNIT)) {
```
Comment (alshan) [inline plot-config-portable/src/commonMain/kotlin/jetbrains/datalore/plot/config/GeomProtoClientSide.kt:122]: Remove 'has' checking: 'getString' returns null if !has()

File: plot-config-portable/src/commonMain/kotlin/jetbrains/datalore/plot/config/GeomProtoClientSide.kt [hunk 12]
```diff
@@ -118,6 +118,19 @@ class GeomProtoClientSide(geomKind: GeomKind) : GeomProto(geomKind) {
...
+                    val sizeUnit = opts.getString(Point.SIZE_UNIT)
+
+                    if ( sizeUnit != null) {
```
Comment (alshan) [inline plot-config-portable/src/commonMain/kotlin/jetbrains/datalore/plot/config/GeomProtoClientSide.kt:125]: May be re-write as: `sizeUnit?.let {}`

File: plot-config-portable/src/commonMain/kotlin/jetbrains/datalore/plot/config/GeomProtoClientSide.kt [hunk 13]
```diff
@@ -118,6 +118,19 @@ class GeomProtoClientSide(geomKind: GeomKind) : GeomProto(geomKind) {
                 if (opts.has(Point.ANIMATION)) {
                     geom.animation = opts[Point.ANIMATION]
                 }
+
+                if (opts.has(Point.SIZE_UNIT)) {
+                    val sizeUnit = opts.getString(Point.SIZE_UNIT)
```
Comment (alshan) [inline plot-config-portable/src/commonMain/kotlin/jetbrains/datalore/plot/config/GeomProtoClientSide.kt:123]: `opts.getString(Point.SIZE_UNIT)?.toLowerCase()`

File: plot-config-portable/src/commonMain/kotlin/jetbrains/datalore/plot/config/GeomProtoClientSide.kt [hunk 14]
```diff
@@ -139,11 +152,24 @@ class GeomProtoClientSide(geomKind: GeomKind) : GeomProto(geomKind) {
 
                     if (naValue != null) {
                         geom.naValue = naValue
-                    }else {
+                    } else {
                         throw IllegalArgumentException("Expected: na_value = 'some string'")
                     }
                 }
 
+                if (opts.has(Text.SIZE_UNIT)) {
```
Comment (alshan) [inline plot-config-portable/src/commonMain/kotlin/jetbrains/datalore/plot/config/GeomProtoClientSide.kt:160]: See comments for Point.SIZE_UNIT

**Review by alshan — CHANGES_REQUESTED** [review]

File: plot-base-portable/src/commonMain/kotlin/jetbrains/datalore/plot/base/geom/TextGeom.kt (hunk 5 above)
Comment (IKrukov-HORIS) [reply plot-base-portable/src/commonMain/kotlin/jetbrains/datalore/plot/base/geom/TextGeom.kt:21]: All of them represent configurable parameters and accessed from GeomProtoClientSide now. Similar way for passing parameters is used in stats. Do you propose to change this protocol?

File: plot-config/src/jvmTest/kotlin/plot/PlotSvgExportTest.kt [hunk 15]
```diff
@@ -33,14 +33,14 @@ internal class PlotSvgExportTest {
...
-    @Test
-    fun svgFromGGBunch() {
-        val svg = PlotSvgExport.buildSvgImageFromRawSpecs(
-            plotSpec = rawSpec_GGBunch(),
-            plotSize = DoubleVector(400.0, 300.0)  // Ignored
-        )
-
-        kotlin.test.assertEquals(EXPECTED_BUNCH_SVG, svg)
-    }
+//
```
[the remark this answers is not in the document]
Comment (IKrukov-HORIS) [reply plot-config/src/jvmTest/kotlin/plot/PlotSvgExportTest.kt:36]: Fixed

File: plot-config-portable/src/commonMain/kotlin/jetbrains/datalore/plot/config/Option.kt [hunk 16]
```diff
@@ -139,6 +139,12 @@ object Option {
 
         object Point {
             const val ANIMATION = "animation"
+
+            object SizeUnit {
```
Comment (alshan) [inline plot-config-portable/src/commonMain/kotlin/jetbrains/datalore/plot/config/Option.kt:143]: Pls. move`SIZE_UNIT` back to `Point`.
Only x, y should be in `SizeUnit`. `SizeUnit` itself must reside in `Geom`.
The access should look like:
Geom.Point.SIZE_UNIT
Geom.SizeUnit.X

File: plot-config/src/jvmTest/kotlin/plot/PlotSvgExportTest.kt (hunk 15 above)
Comment (alshan) [inline plot-config/src/jvmTest/kotlin/plot/PlotSvgExportTest.kt:36]: remove

File: plot-config/src/jvmTest/kotlin/plot/PlotSvgExportTest.kt [hunk 17]
```diff
@@ -33,14 +33,14 @@ internal class PlotSvgExportTest {
...
+//    @Test
```
Comment (alshan) [inline plot-config/src/jvmTest/kotlin/plot/PlotSvgExportTest.kt:37]: uncomment

File: python-package/lets_plot/bistro/corr.py [hunk 18]
```diff
@@ -0,0 +1,383 @@
...
+_PLOT_PROPORTION = 3.0 / 4.0
+
+
+class corr_plot_builder:
+    """
+    This class is intended to build correlation matrix plots.
+    """
+
+    def __init__(self, data, show_legend=None, format=None, flip=None):
+        """
```
Comment (alshan) [inline python-package/lets_plot/bistro/corr.py:76]: We are using [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard) doc-string standard. For example: For example: https://github.com/JetBrains/lets-plot/blob/2efff4cdd371a826a6718b448c1646bfdc14e1b7/python-package/lets_plot/plot/scale.py

Please, update.

File: python-package/lets_plot/bistro/corr.py [hunk 19]
```diff
@@ -0,0 +1,383 @@
...
+
+    def palette_gradient(self, low, mid, high):
+        """
+        Set scale_color_gradient2 for corr plot
+        :param low: color, corresponded to correlation -1
+        :param mid: color, corresponded to correlation 0
+        :param high:color, corresponded to correlation 1
+        :return: self
+        """
+        self._color_scale = scale_color_gradient2(name='Correlation',
```
Comment (alshan) [inline python-package/lets_plot/bistro/corr.py:141]: Also may need to create scale_fill depending on geometries used.
Tiles - fill
Labels - color
Point - color (with the current default shape)

File: plot-config-portable/src/commonMain/kotlin/jetbrains/datalore/plot/config/GeomProtoClientSide.kt [hunk 20]
```diff
@@ -118,6 +118,19 @@ class GeomProtoClientSide(geomKind: GeomKind) : GeomProto(geomKind) {
                 if (opts.has(Point.ANIMATION)) {
                     geom.animation = opts[Point.ANIMATION]
                 }
+
+                val sizeUnit = opts.getString(Point.SizeUnit.SIZE_UNIT)?.toLowerCase()
+                sizeUnit?.let {
+                    if (sizeUnit != Point.SizeUnit.X && sizeUnit != Point.SizeUnit.Y) {
```
Comment (alshan) [inline plot-config-portable/src/commonMain/kotlin/jetbrains/datalore/plot/config/GeomProtoClientSide.kt:124]: We generally do not check values validity here. Let's replace this entire block with:
`geom.sizeUnit = opts.getString(Point.SIZE_UNIT)?.toLowerCase()`

File: plot-config-portable/src/commonMain/kotlin/jetbrains/datalore/plot/config/GeomProtoClientSide.kt [hunk 21]
```diff
@@ -139,11 +152,23 @@ class GeomProtoClientSide(geomKind: GeomKind) : GeomProto(geomKind) {
...
                         geom.naValue = naValue
-                    }else {
+                    } else {
                         throw IllegalArgumentException("Expected: na_value = 'some string'")
                     }
                 }
 
+                val sizeUnit = opts.getString(Text.SizeUnit.SIZE_UNIT)?.toLowerCase()
+                sizeUnit?.let {
+                    if (sizeUnit != Text.SizeUnit.X && sizeUnit != Text.SizeUnit.Y) {
```
Comment (alshan) [inline plot-config-portable/src/commonMain/kotlin/jetbrains/datalore/plot/config/GeomProtoClientSide.kt:162]: Same as Point.

File: plot-base-portable/src/commonMain/kotlin/jetbrains/datalore/plot/base/geom/TextLegendKeyElementFactory.kt [hunk 22]
```diff
@@ -18,7 +18,7 @@ internal class TextLegendKeyElementFactory :
 
     override fun createKeyElement(p: DataPointAesthetics, size: DoubleVector): SvgGElement {
         val label = TextLabel("a")
-        GeomHelper.decorate(label, p)
+        GeomHelper.decorate(label, p,1.0)
```
Comment (alshan) [inline plot-base-portable/src/commonMain/kotlin/jetbrains/datalore/plot/base/geom/TextLegendKeyElementFactory.kt:21]: add space

**Review by alshan — CHANGES_REQUESTED** [review]

File: python-package/lets_plot/bistro/corr.py (hunk 18 above)
Comment (IKrukov-HORIS) [reply python-package/lets_plot/bistro/corr.py:76]: Done

**Pull request closed without merging**